value/value.go

```diff
@@ -17,76 +17,170 @@ limitations under the License.
+	return FromJSONFast(input)
+}
+
+// FromJSONFast is a helper function for reading a JSON document.
+func FromJSONFast(input []byte) (Value, error) {
+	iter := readPool.BorrowIterator(input)
+	defer readPool.ReturnIterator(iter)
+	return ReadJSONIter(iter)
+}
+
+// ToJSON is a helper function for producing a JSon document.
+func ToJSON(v Value) ([]byte, error) {
+	buf := bytes.Buffer{}
+	stream := writePool.BorrowStream(&buf)
+	defer writePool.ReturnStream(stream)
+	WriteJSONStream(v, stream)
+	b := stream.Buffer()
+	err := stream.Flush()
+	// Help jsoniter manage its buffers--without this, the next
+	// use of the stream is likely to require an allocation. Look
+	// at the jsoniter stream code to understand why. They were probably
+	// optimizing for folks using the buffer directly.
+	stream.SetBuffer(b[:0])
+	return buf.Bytes(), err
+}
+
+// ReadJSONIter reads a Value from a JSON iterator.
+func ReadJSONIter(iter *jsoniter.Iterator) (Value, error) {
+	v := iter.Read()
+	if iter.Error != nil && iter.Error != io.EOF {
+		return nil, iter.Error
+	}
+	return NewValueInterface(v), nil
+}
+
+// WriteJSONStream writes a value into a JSON stream.
+func WriteJSONStream(v Value, stream *jsoniter.Stream) {
+	stream.WriteVal(v.Unstructured())
+}
+
+// ToYAML marshals a value as YAML.
+func ToYAML(v Value) ([]byte, error) {
+	return yaml.Marshal(v.Unstructured())
+}
+
+// A Value corresponds to an 'atom' in the schema.
+type Value interface {
+	// IsMap returns true if the Value can be converted to a Map,
+	// false otherwise.
+	IsMap() bool
+	// IsList returns true if the Value can be converted to a List,
+	// false otherwise.
+	IsList() bool
+	// IsBool returns true if the Value can be converted to a bool,
+	// false otherwise.
+	IsBool() bool
+	// IsInt returns true if the Value can be converted to a int64,
+	// false otherwise.
+	IsInt() bool
+	// IsFloat returns true if the Value can be converted to a
+	// float64, false otherwise.
+	IsFloat() bool
+	// IsString returns true if the Value can be converted to a
+	// string, false otherwise.
+	IsString() bool
+	// IsMap returns true if the Value is null, false otherwise.
+	IsNull() bool
+
+	// Map converts the Value into a Map (or panic if the type
+	// doesn't allow it).
+	Map() Map
+	// List converts the Value into a List (or panic if the type
+	// doesn't allow it).
+	List() List
+	// Bool converts the Value into a bool (or panic if the type
+	// doesn't allow it).
+	Bool() bool
+	// Int converts the Value into an int64 (or panic if the type
+	// doesn't allow it).
+	Int() int64
+	// Float converts the Value into a float64 (or panic if the type
+	// doesn't allow it).
+	Float() float64
+	// String converts the Value into a string (or panic if the type
+	// doesn't allow it).
+	String() string
+
+	// Returns a value of this type that is no longer needed. The
+	// value shouldn't be used after this call.
+	Recycle()
+
+	// Converts the Value into an Unstructured interface{}.
```

Comment:
We'll probably want to change this to make the reflect stuff optimal.

I think a `NewOfSameType() Value` might come in handy, as well as a `Set(valueToCopy Value)`. (followup)

Reply:
Absolutely agree!